text/0000-or-patterns.md

```diff
@@ -0,0 +1,666 @@
+from our complexity budget.
+
+With this RFC, we try to reduce the complexity of the language by extending
+a feature which already exists, and which many users already know about,
+to another place. In a sense, giving the user more capabilities results
+in a negative increase in complexity.
+
+[RFC 2175]: https://github.com/rust-lang/rfcs/pull/2175
+
+In concrete terms, where before we only allowed a pattern of the form
+`'|'? pat | pat` at the top level of `match` and [similar constructs][RFC 2175],
+which special cased the language, we now allow `pat | pat` anywhere a pattern
+may occur whereby we simplify the ruleset of the language.
+In fact, there are already users that try this expecting it to work but
+then find out that it does not.
+
+Furthermore, allowing `pat | pat` in the pattern grammar also allows macros to
+produce disjunctions such as `$p | $q`.
+
+## Real world use cases
+
+This RFC wouldn't be complete without concrete use cases which it would
+facilitate. While there are not an overabundance of cases where `pat | pat`
+would help, there are some where it would. Let's go through a few of them.
+
+[precursor]: https://github.com/rust-lang/rfcs/blob/de235887a80555427314c7eb25c6214523d50cce/text/0000-pipe-in-patterns.md
+
+1. One example which was raised in the [precursor] to this RFC was building a
+   state machine which is iterating through `chars_indices`:
+
+   ```rust
+   match iter.next() {
+       Some(_, ' ' | '\n' | '\r' | '\u{21A1}') => {
+           // Change state
+       }
+       Some(index, ch) => {
+           // Look at char
+       }
+       None => return Err(Eof),
+   }
+   ```
+
+[GHC proposal]: https://github.com/osa1/ghc-proposals/blob/or_patterns/proposals/0000-or-patterns.rst#real-world-examples
+
+2. Other examples are listed in the equivalent [GHC proposal].
+
+3. Another example which was provided in the [precursor] RFC was:
+
+   ```rust
+   for event in event_pump.poll_iter() {
+       use sdl2::event::Event;
+       use sdl2::keyboard::Keycode::{Escape, Q};
+       match event {
+           Event::KeyDown { keycode: Some(Escape | Q), ... } => break 'game,
+           _ => {},
+       }
+       ...
+   }
+   ```
+
+4. Other cases where this feature was requested include:
+   + <https://github.com/rust-lang/rust/issues/15219>
+   + <https://github.com/rust-lang/rust/issues/14516>
+
+[alercah_discord]: https://discordapp.com/channels/442252698964721669/448237931136679936/483325957130813440
+
+5. Another use case due to [@alercah][alercah_discord] is:
+
+   ```rust
+   pub fn is_green(self) -> bool {
+       match self {
+           | Tile::Suited(Suit::Souzu, 2 | 3 | 4 | 6 | 8)
+           | Tile::Dragon(Dragon::Green) => true,
+           _ => false,
+       }
+   }
+   ```
+
+6. Some further examples found with sourcegraph include:
+
+   From [cc-rs](https://sourcegraph.com/github.com/alexcrichton/cc-rs/-/blob/src/lib.rs#L1307:18):
```

Comment:
For all of these citation links, please link to a specific commit hash, so that the line number references remain working in the future.